### ISSUE TITLE
Support for parameter profiles and fields.

### ISSUE THREAD

**Issue by ali-ramadhan** [CliMA/Oceananigans.jl#51]
So instead of having, for example, a constant eddy diffusivity κ, it can be a profile κ(z) or a field κ(x, y, z), or even κ(x, y, z, t).

**Comment by ali-ramadhan** [comment]
Some useful ideas from @peterahrens maybe `ConstantField`, `ProfileField`, etc.
```julia
type AbstractField{T, D} end

struct ConstantField{T, D}
    val::T
end

get_value(f::ConstantField{T, D}, x:::NTuple{D}) where {T, D} = f.val


struct DeformedField{T, D, X::Field{NTuple{D}, D}, F<:Field{T, D}}
    ξ::X
    parent::F
end

get_value(f::DeformedField{T, D}, x::Vararg{Any, D}) where {T, D} = get_value(f.parent, f.ξ(x))

function Base.:+(a::DeformedField, b::DeformedField)
    @assert a.ξ == b.ξ
    DeformedField(a.ξ, a.parent + b.parent)
end

function Base.:+(a::DeformedField, s::Number)
    DeformedField(a.ξ, a.parent * s)
end
```

**Comment by glwagner** [comment]
It's easy. The main challenge is designing the API / semantics. Do we want to extend `ConstantIsotropicDiffusivity` to new `κ` types (dispatching on the type of `κ` to achieve this goal), or do we want to create a new type, eg `SpatiallyVaryingIsotropicDiffusivity`?

If we nudge the meaning of "constant" to mean "constant in time", I think that it could make the most sense to extend the functionality of `ConstantIsotropicDiffusivity`.

Also, I am realizing that we want to do this with functions rather than arrays. That's ok, because we can also compute flux divergences for diffusivity functions.

**Comment by ali-ramadhan** [comment]
I think a new type would reduce confusion.

Right now `ConstantIsotropicDiffusivity` is fine as it's both constant in time and space, but could be confusing if it's just constant in time. `SpatiallyVaryingIsotropicDiffusivity` is a bit of a mouthful, but being verbose is probably better than having the user guess or have them continuously checking the documentation.

> Also, I am realizing that we want to do this with functions rather than arrays.

Arrays can always be implemented via a function that accesses a `const` array anyways, but would be good to benchmark.

**Comment by ali-ramadhan** [comment]
This has been resolved by PR #806.
